In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
# import the “products.csv” data set 
df_prods = pd.read_csv(r'C:\Users\ZenBook\Instacart Basket Analysis_30.09.2025\02 Data\Original Data\products.csv', index_col = False)

In [4]:
# import “orders_wrangled.csv” data set
df_ords = pd.read_csv(r'C:\Users\ZenBook\Instacart Basket Analysis_30.09.2025\02 Data\Prepared Data\orders_wrangled.csv', index_col = False)

# Data Consistency

# Mixed-Type Data

In [5]:
# create a dataframe
df_test = pd.DataFrame()

In [6]:
# create a mixed-type column 
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
# convert a column’s data type from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

In [10]:
# Finding Missing Values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
# create a subset of the dataframe 'product_name'
df_nan = df_prods[df_prods['product_name'].isnull () == True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# Addressing Missing Values

In [13]:
df_prods.shape

(49693, 5)

In [14]:
# create a new dataframe called df_prods_clean with non-missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [18]:
df_prods_clean.shape

(49677, 5)

In [15]:
# or delete missing values
df_prods.dropna(subset = ['product_name'], inplace = True)

# Duplicates

In [16]:
# Finding Duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [17]:
# show new dataframe that consists only duplicates
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [ ]:
# Deletion Duplicates

In [18]:
# current number of rows in df_prods_clean dataframe 
df_prods_clean.shape

(49677, 5)

In [19]:
# created a new dataframe without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [20]:
df_prods_clean_no_dups.shape

(49672, 5)

# ANSWERS FOR A TASK

# 2. Run the df.describe() function on your df_ords dataframe

In [21]:
# import the “orders.csv” data set 
df_ords = pd.read_csv(r'C:\Users\ZenBook\Instacart Basket Analysis_30.09.2025\02 Data\Original Data\orders.csv', index_col = False)

In [22]:
pd.set_option('display.float_format', '{:.0f}'.format)

In [23]:
df_ords.describe()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
count   3421083  3421083       3421083    3421083            3421083   
mean    1710542   102978            17          3                 13   
std      987582    59534            18          2                  4   
min           1        1             1          0                  0   
25%      855272    51394             5          1                 10   
50%     1710542   102689            11          3                 13   
75%     2565812   154385            23          5                 16   
max     3421083   206209           100          6                 23   

       days_since_prior_order  
count                 3214874  
mean                       11  
std                         9  
min                         0  
25%                         4  
50%                         7  
75%                        15  
max                        30

Whether anything about the data looks off or should be investigated further?
- the max order_number is 100
- order_dow (day of week): 0 represents Monday,..., 6 represents Sunday, while it should be 7.
- count of the days_since_prior_order is differ from values in other columns (3214874 instead 3421083) suggesting missing data

In [24]:
df_prods.describe()

product_id  aisle_id  department_id  prices
count       49677     49677          49677   49677
mean        24850        68             12      10
std         14341        38              6     454
min             1         1              1       1
25%         12433        35              7       4
50%         24851        69             13       7
75%         37267       100             17      11
max         49688       134             21   99999

Whether anything about the data 'df_prods' looks off or should be investigated further?
There is extreme outlier in prices - Max = 99999 while the mean = 10 and 75% = 11. 
This suggests a likely data entry error or placeholder value. It will be recommendet to check if 99999 is used as a missing value proxy or misrecorded.

# 3. Check for mixed-type data in your df_ords dataframe.

In [25]:
# check for mixed-type data
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

There are no mixed data-types in the df_ords

In [26]:
df_ords.dtypes

order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
order_dow                   int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

# 5. Check for missing values in your df_ords dataframe

In [27]:
df_ords.shape

(3421083, 7)

In [28]:
# finding missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206209 cells with missing values in the days_since_prior_order column. It is likely correspond to users' first-ever orders, where no previous purchase exists to calculate a time gap. 

# 6. Address the missing values

In [38]:
df_ords.shape

(3421083, 7)

In [29]:
# create a subset of the dataframe 'days_since_prior_order'
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull () == True]

In [30]:
df_ords_nan

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

Seems like the missing values in the column days_since_prior_order are customer's first order. So these records cannot be deleted. Instead we can introduce new column which clarify it is true or not.

In [31]:
# Identify first-time orders with no prior purchase history
df_ords['first_order'] = df_ords['days_since_prior_order'].isna()

In [32]:
df_ords

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  first_order  
0                        8                     NaN         True  
1                        7                      15        False  
2                       12                      21        False  
3                        7                      29        False  
4                       15                      28        False  
...                    ...                     ...          ...  
3421078                 18                      29        False  
3421079                 10                      30        False  
3421080                 12                      18        False  
3421081                 12                       7        False  
3421082                 14                      30        False  

[3421083 rows x 8 columns]

In [33]:
# creating a dataframe 'df_ords_first_order', where order_number will be 1
df_ords_first_order = df_ords[df_ords['order_number'] == 1]

In [34]:
df_ords_first_order

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  first_order  
0                        8                     NaN         True  
11                      11                     NaN         True  
26                      14                     NaN         True  
39                      11                     NaN         True  
45                      12                     NaN         True  
...                    ...                     ...          ...  
3420930                 12                     NaN         True  
3420934                 18                     NaN         True  
3421002                 19                     NaN         True  
3421019                 15                     NaN         True  
3421069                 11                     NaN         True  

[206209 rows x 8 columns]

So, all first_order values are marked as a True. 

# 7. Run a check for duplicate values in your df_ords data.

In [35]:
# finding duplicates in the df_ords
df_ords_dups = df_ords[df_ords.duplicated()]

In [36]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order, first_order]
Index: []

No duplicates founded in the df_ords.

# 8. Address the duplicates using an appropriate method.

In [37]:
# current number of rows in df_orders dataframe 
df_ords.shape

(3421083, 8)

In [38]:
# created a new dataframe without duplicates
df_ords_no_dups = df_ords.drop_duplicates()

In [39]:
df_ords_no_dups

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  first_order  
0                        8                     NaN         True  
1                        7                      15        False  
2                       12                      21        False  
3                        7                      29        False  
4                       15                      28        False  
...                    ...                     ...          ...  
3421078                 18                      29        False  
3421079                 10                      30        False  
3421080                 12                      18        False  
3421081                 12                       7        False  
3421082                 14                      30        False  

[3421083 rows x 8 columns]

No duplicates - no change in data shape.

# Export final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder 

In [40]:
df_prods.to_csv(r'C:\Users\ZenBook\Instacart Basket Analysis_30.09.2025\02 Data\Prepared Data\products_checked.csv', index=False)

In [41]:
df_ords.to_csv(r'C:\Users\ZenBook\Instacart Basket Analysis_30.09.2025\02 Data\Prepared Data\orders_checked.csv', index=False)